src/buildstream/storage/_casbaseddirectory.py

```diff
@@ -114,136 +128,337 @@ def get_equivalency_properties(e: IndexEntry):
+    #############################################################
 
-            self.index[entry.name] = IndexEntry(
-                entry.name,
-                _FileType.REGULAR_FILE,
-                digest=entry.digest,
-                is_executable=entry.is_executable,
-                mtime=mtime,
+    def descend(self, path: str, *, create: bool = False, follow_symlinks: bool = False) -> "CasBasedDirectory":
+        self._validate_path(path)
+        paths = path.split("/")
+        return self.__descend(paths, create=create, follow_symlinks=follow_symlinks)
+
+    def import_single_file(self, external_pathspec: str) -> FileListResult:
+        result = FileListResult()
+        if self.__check_replacement(os.path.basename(external_pathspec), os.path.dirname(external_pathspec), result):
+            self.__add_file(
+                os.path.basename(external_pathspec), external_pathspec, properties=None,
             )
-        for entry in pb2_directory.symlinks:
-            self.index[entry.name] = IndexEntry(entry.name, _FileType.SYMLINK, target=entry.target)
+            result.files_written.append(external_pathspec)
+        return result
+
+    def export_to_tar(self, tarfile: TarFile, destination_dir: str, mtime: int = BST_ARBITRARY_TIMESTAMP) -> None:
+        for filename, entry in sorted(self.__index.items()):
+            arcname = os.path.join(destination_dir, filename)
+            if entry.type == FileType.DIRECTORY:
+                tarinfo = tarfilelib.TarInfo(arcname)
+                tarinfo.mtime = mtime
+                tarinfo.type = tarfilelib.DIRTYPE
+                tarinfo.mode = 0o755
+                tarfile.addfile(tarinfo)
+                self.descend(filename).export_to_tar(tarfile, arcname, mtime)
+            elif entry.type == FileType.REGULAR_FILE:
+                source_name = self.__cas_cache.objpath(entry.digest)
+                tarinfo = tarfilelib.TarInfo(arcname)
+                tarinfo.mtime = mtime
+                if entry.is_executable:
+                    tarinfo.mode |= stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH
+                tarinfo.size = os.path.getsize(source_name)
+                with open(source_name, "rb") as f:
+                    tarfile.addfile(tarinfo, f)
+            elif entry.type == FileType.SYMLINK:
+                assert entry.target is not None
+                tarinfo = tarfilelib.TarInfo(arcname)
+                tarinfo.mtime = mtime
+                tarinfo.mode = 0o777
+                tarinfo.linkname = entry.target
+                tarinfo.type = tarfilelib.SYMTYPE
+                sio = StringIO(entry.target)
+                bio = BytesIO(sio.read().encode("utf8"))
+                tarfile.addfile(tarinfo, bio)
+            else:
+                raise DirectoryError("can not export file type {} to tar".format(entry.type))
+
+    def is_empty(self) -> bool:
+        return len(self.__index) == 0
+
+    def listdir(self) -> Iterator[str]:
```

Comment:
This is not declared in `Directory`. I'm wondering whether we should combine this and `list_relative_paths()` into a single method with an optional `recursive` (bool keyword-only) parameter. What do you think?

Reply:
oops listdir() is an artifact of something I've already backed out of the branch.

This is handled by way of the `Directory` being iterable instead, I'll remove this from wherever it was added.